protocols/gossipsub/src/protocol.rs

```diff
@@ -267,37 +267,28 @@ impl Encoder for GossipsubCodec {
     type Item = rpc_proto::Rpc;
     type Error = GossipsubHandlerError;
 
-    fn encode(&mut self, item: Self::Item, dst: &mut BytesMut) -> Result<(), Self::Error> {
-        let mut buf = Vec::with_capacity(item.encoded_len());
-
-        item.encode(&mut buf)
-            .expect("Buffer has sufficient capacity");
-
-        // length prefix the protobuf message, ensuring the max limit is not hit
-        self.length_codec
-            .encode(Bytes::from(buf), dst)
-            .map_err(|_| GossipsubHandlerError::MaxTransmissionSize)
+    fn encode(
+        &mut self,
+        item: Self::Item,
+        dst: &mut BytesMut,
+    ) -> Result<(), GossipsubHandlerError> {
+        match self.codec.encode(item, dst) {
+            Ok(_) => Ok(()),
+            _ => Err(GossipsubHandlerError::MaxTransmissionSize),
+        }
     }
 }
 
 impl Decoder for GossipsubCodec {
     type Item = HandlerEvent;
     type Error = GossipsubHandlerError;
 
-    fn decode(&mut self, src: &mut BytesMut) -> Result<Option<Self::Item>, Self::Error> {
-        let packet = match self.length_codec.decode(src).map_err(|e| {
-            if let std::io::ErrorKind::PermissionDenied = e.kind() {
-                GossipsubHandlerError::MaxTransmissionSize
-            } else {
-                GossipsubHandlerError::Io(e)
-            }
-        })? {
-            Some(p) => p,
-            None => return Ok(None),
+    fn decode(&mut self, src: &mut BytesMut) -> Result<Option<Self::Item>, GossipsubHandlerError> {
+        let rpc = match self.codec.decode(src) {
```

Comment:
Same here, please use `?`.

Reply:
Hmmm, when I have `let rpc = self.codec.decode(src)?.unwrap();` the smoke tests fail with this message:

```
thread 'libp2p-swarm-task-8' panicked at 'called `Option::unwrap()` on a `None` value', protocols/gossipsub/src/protocol.rs:284:43
```

I've tried variations on `unwrap()`, but none seem be able to `return Ok(None)`.

I've settled on this code for these reasons:
1) If the old code saw a PermissionDenied error, it was setting the error to `GossipsubHandlerError::MaxTransmissionSize` which doesn't seem right.
2) Although it isn't using the `?`, it does seem to cover the cases and return None (which allows the smoke tests to pass)